src/memray/_memray/snapshot.cpp

```diff
@@ -122,6 +122,77 @@ SnapshotAllocationAggregator::getSnapshotAllocations(bool merge_threads)
     return stack_to_allocation;
 }
 
+TemporaryAllocationsAggregator::TemporaryAllocationsAggregator(size_t max_items)
+: d_max_items(max_items)
+{
+}
+
+void
+TemporaryAllocationsAggregator::addAllocation(const Allocation& allocation)
+{
+    hooks::AllocatorKind kind = hooks::allocatorKind(allocation.allocator);
+    auto it = d_current_allocations.find(allocation.tid);
+    switch (kind) {
+        case hooks::AllocatorKind::SIMPLE_ALLOCATOR: {
+            case hooks::AllocatorKind::RANGED_ALLOCATOR: {
+                if (it == d_current_allocations.end()) {
+                    it = d_current_allocations.insert(
+                            it,
+                            std::pair(allocation.tid, std::deque<Allocation>()));
+                }
+
+                it->second.emplace_front(allocation);
+                break;
+            }
+            case hooks::AllocatorKind::SIMPLE_DEALLOCATOR:
+            case hooks::AllocatorKind::RANGED_DEALLOCATOR: {
+                if (it == d_current_allocations.end()) {
+                    break;
+                }
+
+                auto alloc_it = std::find_if(
+                        it->second.begin(),
+                        it->second.end(),
+                        [&](auto& current_allocation) {
+                            bool match = (current_allocation.address == allocation.address);
```

Comment:
I'm not sure there's anything we can reasonably do about it, but this could have false positives if pointers are passed between threads and freed by some other thread, due to pointers being pooled and reused.

Imagine pointer 0x1234 gets allocated in thread A, and passed to thread B on a `queue.Queue`. Thread B does some processing of that work item and then destroys it. Thread B then creates a new object and gets pointer 0x1234 again, reusing it for the new object. Thread B passes that object back to thread A as a response. Thread A logs the response and destroys it.

We'll think that deallocation of 0x1234 is for the first object, which was allocated on thread A, and we'll incorrectly decide that the first allocation was temporary.

I'm not sure there's any good solution for that, though, without ignoring thread IDs entirely and mashing all allocations from all threads together.

Reply:
Yeah, I though about this but there isn't anything that we can do here. The other alternative is flatten all threads but that not only is more racy but will cause more problems than what you mention as many threads racing allocations will interleave the allocs/dealloc so we won't detect them correctly.